File: dotnet/Core/Workspace/CSharp/configuration/core/workspace/derivations/default/Engine.cs
// <copyright file="DerivationService.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Configuration.Derivations.Default
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Meta;
    using Allors.Workspace.Derivations;

    public class Engine
    {
        public IDictionary<IRule, ISet<IClass>> ClassesByRule { get; }

        public IDictionary<IClass, IRule[]> RulesByClass { get; }

        public IDictionary<IClass, IDictionary<IRoleType, ISet<IRolePattern>>> PatternsByRoleTypeByClass { get; }

        public IDictionary<IClass, IDictionary<IAssociationType, ISet<IAssociationPattern>>> PatternsByAssociationTypeByClass { get; }

        public IDictionary<IPattern, IRule> RuleByPattern { get; }

        public Engine(IRule[] rules)
        {
            this.ClassesByRule = new Dictionary<IRule, ISet<IClass>>();
            this.PatternsByRoleTypeByClass = new Dictionary<IClass, IDictionary<IRoleType, ISet<IRolePattern>>>();
            this.PatternsByAssociationTypeByClass = new Dictionary<IClass, IDictionary<IAssociationType, ISet<IAssociationPattern>>>();
            this.RuleByPattern = new Dictionary<IPattern, IRule>();

            foreach (var rule in rules)
            {
                var ruleClasses = new HashSet<IClass>();
                foreach (var pattern in rule.Patterns)
                {
                    this.RuleByPattern.Add(pattern, rule);

                    var patternClasses = pattern switch
                    {
                        IRolePattern { ObjectType: null } rolePattern => rolePattern.RoleType.AssociationType.ObjectType.Classes.ToArray(),
                        IRolePattern { ObjectType: { } } rolePattern => rolePattern.ObjectType.Classes.ToArray(),
                        IAssociationPattern associationPattern => associationPattern.AssociationType.RoleType.ObjectType.IsComposite ? ((IComposite)associationPattern.AssociationType.RoleType.ObjectType).Classes.ToArray() : Array.Empty<IClass>(),
                        _ => Array.Empty<IClass>()
                    };

                    ruleClasses.UnionWith(patternClasses);

                    switch (pattern)
                    {
                        case IRolePattern rolePattern:
                            foreach (var @class in patternClasses)
                            {
                                if (!this.PatternsByRoleTypeByClass.TryGetValue(@class, out var patternsByRoleType))
                                {
                                    patternsByRoleType = new Dictionary<IRoleType, ISet<IRolePattern>>();
                                    this.PatternsByRoleTypeByClass.Add(@class, patternsByRoleType);
                                }

                                var roleType = rolePattern.RoleType;

                                if (!patternsByRoleType.TryGetValue(roleType, out var patterns))
                                {
                                    patterns = new HashSet<IRolePattern>();
                                    patternsByRoleType.Add(roleType, patterns);
                                }

                                patterns.Add(rolePattern);
                            }

                            break;
                        case IAssociationPattern associationPattern:
                            foreach (var @class in patternClasses)
                            {
                                if (!this.PatternsByAssociationTypeByClass.TryGetValue(@class, out var patternsByAssociationType))
                                {
                                    patternsByAssociationType = new Dictionary<IAssociationType, ISet<IAssociationPattern>>();
                                    this.PatternsByAssociationTypeByClass.Add(@class, patternsByAssociationType);
                                }

                                var associationType = associationPattern.AssociationType;

                                if (!patternsByAssociationType.TryGetValue(associationType, out var patterns))
                                {
                                    patterns = new HashSet<IAssociationPattern>();
                                    patternsByAssociationType.Add(associationType, patterns);
                                }

                                patterns.Add(associationPattern);
                            }

                            break;
                    }
                }

                this.ClassesByRule.Add(rule, ruleClasses);
            }

            var classes = this.ClassesByRule.SelectMany(v => v.Value).Distinct();
            this.RulesByClass = classes.ToDictionary(v => v, v => rules.Where(w => this.ClassesByRule[w].Contains(v)).ToArray());
        }
    }
}

File: typescript/libs/workspace/configuration/core/src/lib/derivations/default/Engine.ts
import { IAssociationPattern, IPattern, IPatternKind, IRolePattern, IRule } from '@allors/workspace/domain/system';
import { AssociationType, Class, Composite, RoleType } from '@allors/workspace/meta/system';

export class Engine {
  classesByRule: Map<IRule, Set<Class>>;

  rulesByClass: Map<Class, IRule[]>;

  patternsByRoleTypeByClass: Map<Class, Map<RoleType, Set<IRolePattern>>>;

  patternsByAssociationTypeByClass: Map<Class, Map<AssociationType, Set<IAssociationPattern>>>;

  ruleByPattern: Map<IPattern, IRule>;

  public constructor(rules: IRule[]) {
    this.classesByRule = new Map();
    this.patternsByRoleTypeByClass = new Map();
    this.patternsByAssociationTypeByClass = new Map();
    this.ruleByPattern = new Map();

    for (const rule of rules) {
      let ruleClasses = new Set<Class>();

      for (const pattern of rule.patterns) {
        this.ruleByPattern.set(pattern, rule);

        let patternClasses: Set<Class>;

        switch (pattern.kind) {
          case 'RolePattern':
            if (pattern.objectType) {
              patternClasses = pattern.objectType.classes;
            } else {
              patternClasses = (pattern.roleType.associationType.objectType as Composite).classes;
            }
            break;
          case 'AssociationPattern':
            if (pattern.AssociationType.roleType.objectType.isComposite) {
              ruleClasses = new Set([...ruleClasses, ...(pattern.AssociationType.roleType.objectType as Composite).classes]);
            }
            break;
        }

        if (patternClasses) {
          ruleClasses = new Set([...ruleClasses, ...patternClasses]);

          switch (pattern.kind) {
            case 'RolePattern':
              // for (const cls in patternClasses) {
              //     if (!this.PatternsByRoleTypeByClass.TryGetValue(class, /* out */var, patternsByRoleType)) {
              //         patternsByRoleType = new Dictionary<IRoleType, ISet<IRolePattern>>();
              //         this.PatternsByRoleTypeByClass.Add(class, patternsByRoleType);
              //     }

              //     let roleType = rolePattern.RoleType;
              //     if (!patternsByRoleType.TryGetValue(roleType, /* out */var, patterns)) {
              //         patterns = new HashSet<IRolePattern>();
              //         patternsByRoleType.Add(roleType, patterns);
              //     }

              //     patterns.Add(rolePattern);
              // }

              break;
            case 'AssociationPattern':
              // for (let class in patternClasses) {
              //     if (!this.PatternsByAssociationTypeByClass.TryGetValue(class, /* out */var, patternsByAssociationType)) {
              //         patternsByAssociationType = new Dictionary<IAssociationType, ISet<IAssociationPattern>>();
              //         this.PatternsByAssociationTypeByClass.Add(class, patternsByAssociationType);
              //     }

              //     let associationType = associationPattern.AssociationType;
              //     if (!patternsByAssociationType.TryGetValue(associationType, /* out */var, patterns)) {
              //         patterns = new HashSet<IAssociationPattern>();
              //         patternsByAssociationType.Add(associationType, patterns);
              //     }

              //     patterns.Add(associationPattern);
              // }

              break;
          }
        }
      }

      this.classesByRule.set(rule, ruleClasses);
    }

    // var classes = this.classesByRule.SelectMany(v => v.Value).Distinct();
    // this.rulesByClass = classes.ToDictionary(v => v, v => rules.Where(w => this.ClassesByRule[w].Contains(v)).ToArray());
  }
}
